fflib/src/classes/fflib_QueryFactory.cls

```diff
@@ -84,17 +84,51 @@ public class fflib_QueryFactory { //No explicit sharing declaration - inherit fr
 	private Schema.ChildRelationship relationship;
 	private Map<Schema.ChildRelationship, fflib_QueryFactory> subselectQueryMap;
 
-	private String getFieldPath(String fieldName){
-		if(!fieldName.contains('.')){ //single field
+	/**
+	 * Pattern variable to prevent the Pattern to be compiled for each field
+	 */
+	private Pattern soqlFunctionPattern { get{
+		if( soqlFunctionPattern == null ){
+			soqlFunctionPattern = Pattern.compile( '^([a-zA-Z_]*\\(\\s*)([a-zA-Z0-9_\\.]+)?(\\s*\\)[a-zA-Z ]*)$' );
+		}
+		return soqlFunctionPattern;
+	} set; }
+
+	private String getFieldPath( String fieldName ){
```

Comment:
By adding this logic inside this method the complexity suddenly becomes very high. I would advise to split this logic into a number of private methods to keep a level of abstraction here in his method.

Reply:
@wimvelzeboer I do agree with you on the increased complexity and length of the method. I've thought of this before, but in my perspective this will only become more difficult to read, since beforeField and afterField is used throughout the function. The methods should then return a list or an innerclass to support. Global is also not an option since there will be multiple field paths called. That's why I've put emphasize on the comments.

If you have any suggestion how to split and make sure all data is properly fed back to be used, please let me know!

Reply:
I would maybe split up this method into three parts:
`private String getFieldPathForMethod(String fieldName)`
`private String getFieldPathForRelatedField(String fieldName)`
`private String getFieldPathForField(String fieldName)`

```
private String getFieldPath( String fieldName )
{
  if (fieldName.contains( '(' )) return getFieldPathForMethod(fieldName);
  if (fieldName.contains( '.' ) ) return getFieldPathForRelatedField(fieldName);
  return getFieldPathForField(fieldName);
}

private getFieldPathForMethod(String methodCall)
{
  Matcher match = this.soqlFunctionPattern.matcher( methodCall );
  if (!match.matches())
  {
    throw new InvalidFieldException( 'The method '+ methodCall + 'does specify a ( function opening, but doesn\'t match the pattern' );
  }
  
  String beforeField = match.group( 1 );
  String fieldName = match.group( 2 );
  String afterField = match.group( 3 );

  // When no fieldName is specified (e.g. COUNT() ), return directly, since nothing to check
  if( String.isBlank( fieldName ) )
  {
    return beforeField + afterField;
  }
  return beforeField + getFieldPath(fieldName) + afterField;
}
```